src_analysis/fuzzer_dependencies.py

```diff
@@ -0,0 +1,182 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Module for finding dependencies of fuzzers, and fuzzers that are
+dependent on given files.
+This module assumes that a fuzzer module's imports are done in a sane,
+normal way. It will not work on non-toplevel imports.
+The following style of imports are supported:
+1. from fuzzers.afl import fuzzer
+2. from fuzzers.afl import fuzzer as afl_fuzzer
+
+The following are not supported because they will be considered builtin modules.
+1. import fuzzers.afl.fuzzer
+2. import fuzzers.afl
+3. import blah  # Relative-import (against style guide anyway).
+
+This case is not supported because the dependency will not be recognized:
+from fuzzers.afl.fuzzer import build
+
+"""
+import importlib
+import inspect
+import os
+import types
+from typing import Dict, List, Set
+import sys
+
+from common import fuzzer_utils
+
+# The max depth of dependencies _get_python_dependencies will search.
+PY_DEPENDENCIES_MAX_DEPTH = 10
+
+# A cache of Python dependencies for modules. Keys are strings of module paths.
+# Values are sets of module paths.
+PY_DEPENDENCIES_CACHE: Dict[str, Set[str]] = {}
+
+# Cache these so we don't need to do it every time we call
+# get_underlying_fuzzer.
+FUZZER_CONFIGS = fuzzer_utils.get_fuzzer_configs()
+FUZZER_NAMES_TO_UNDERLYING = {
+    fuzzer_utils.get_fuzzer_from_config(config): config['fuzzer']
+    for config in FUZZER_CONFIGS
+}
+
+
+def _get_fuzzer_module_name(fuzzer: str) -> str:
+    """Returns the name of the fuzzer.py module of |fuzzer|. Assumes |fuzzer| is
+    an underlying fuzzer."""
+    return 'fuzzers.{}.fuzzer'.format(fuzzer)
+
+
+def is_builtin_module(module: types.ModuleType) -> bool:
+    """Returns True if |module| is a python builtin module."""
+    return module.__name__ in sys.builtin_module_names
+
+
+def is_fuzzers_subpath(path: str) -> bool:
+    """Returns True if path is a subpath of the fuzzers/ directory."""
+    common_path = os.path.commonpath([path, fuzzer_utils.FUZZERS_DIR])
```

Comment:
os.path.commonpath use to detect if something is part of that directory, can you make a helper function in common/filesystem or something.

Reply:
Done